src/hotspot/share/opto/loopnode.cpp

```diff
@@ -4765,6 +4765,23 @@ void PhaseIdealLoop::eliminate_useless_zero_trip_guard() {
       Node* opaque = head->is_canonical_loop_entry();
       if (opaque != nullptr) {
         useful_zero_trip_guard_opaques_nodes.push(opaque);
+#ifdef ASSERT
+        // See PhaseIdealLoop::do_unroll
+        // This property is desirable, but it maybe not hold after cloning a loop.
+        // In such a case, we bail out from unrolling, and rely on IGVN to cleanup stuff.
+        // We are here before loop cloning (before iteration_split), so if there is a
+        // mess, it must come from the previous round of loop optimizations, which is bad.
+        // On the other hand, if this assert passes, bailing out in do_unroll means that
+        // this property was broken in the current round of loop optimization, which is
+        // acceptable.
```

Comment:
This comment would be more useful if it was a bit more precise: could you clarify (in the comment) what do you mean by "desirable", "stuff", "mess", and "bad"?